core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleReader.scala

```diff
@@ -33,17 +34,54 @@ private[spark] class HashShuffleReader[K, C](
     "Hash shuffle currently only supports fetching one partition")
 
   private val dep = handle.dependency
+  private val blockManager = SparkEnv.get.blockManager
 
   /** Read the combined key-values for this reduce task */
   override def read(): Iterator[Product2[K, C]] = {
+    val blockStreams = BlockStoreShuffleFetcher.fetchBlockStreams(
+      handle.shuffleId, startPartition, context)
+
+    // Wrap the streams for compression based on configuration
+    val wrappedStreams = blockStreams.map { case (blockId, inputStream) =>
+      blockManager.wrapForCompression(blockId, inputStream)
+    }
+
     val ser = Serializer.getSerializer(dep.serializer)
-    val iter = BlockStoreShuffleFetcher.fetch(handle.shuffleId, startPartition, context, ser)
+    val serializerInstance = ser.newInstance()
+
+    // Create a key/value iterator for each stream
+    val recordIterator = wrappedStreams.flatMap { wrappedStream =>
+      val kvIter = serializerInstance.deserializeStream(wrappedStream).asKeyValueIterator
+      CompletionIterator[(Any, Any), Iterator[(Any, Any)]](kvIter, {
+        // Close the stream once all the records have been read from it to free underlying
+        // ManagedBuffer as soon as possible. Note that in case of task failure, the task's
+        // TaskCompletionListener will make sure this is released.
+        wrappedStream.close()
```

Comment:
It looks like "asKeyValueIterator" creates a NextIterator that calls close on the underlying deserialization stream, and at least the Java deserialization stream calls close() on the underlying input stream: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/serializer/JavaSerializer.scala#L70. Is the close() call here somehow called earlier than that (is that what your comment on line 56 is getting at)?